chain-indexer/src/infra/storage/postgres.rs

```diff
@@ -169,6 +173,409 @@ impl Storage for PostgresStorage {
+        &self,
+        generation_info: &[DustGenerationInfo],
+    ) -> Result<(), sqlx::Error> {
+        if generation_info.is_empty() {
+            return Ok(());
+        }
+
+        let mut tx = self.pool.begin().await?;
+
+        let query = indoc! {"
+            INSERT INTO dust_generation_info (
+                value,
+                owner,
+                nonce,
+                ctime,
+                dtime,
+                merkle_index
+            )
+        "};
+
+        QueryBuilder::new(query)
+            .push_values(generation_info.iter(), |mut q, info| {
+                q.push_bind(U128BeBytes::from(info.value))
+                    .push_bind(info.owner.as_ref())
+                    .push_bind(info.nonce.as_ref())
+                    .push_bind(info.ctime as i64)
+                    .push_bind(if info.dtime == 0 {
+                        None
+                    } else {
+                        Some(info.dtime as i64)
+                    })
+                    .push_bind(0i64); // TODO: merkle_index should come from somewhere.
+            })
+            .build()
+            .execute(&mut *tx)
+            .await?;
+
+        tx.commit().await
+    }
+
+    async fn save_cnight_registrations(
+        &self,
+        registrations: &[DustRegistration],
+    ) -> Result<(), sqlx::Error> {
+        if registrations.is_empty() {
+            return Ok(());
+        }
+
+        let mut tx = self.pool.begin().await?;
+
+        let query = indoc! {"
+            INSERT INTO cnight_registrations (
+                cardano_address,
+                dust_address,
+                is_valid,
+                registered_at,
+                removed_at
+            )
+        "};
+
+        QueryBuilder::new(query)
+            .push_values(registrations.iter(), |mut q, reg| {
+                q.push_bind(reg.cardano_address.as_ref())
+                    .push_bind(reg.dust_address.as_ref())
+                    .push_bind(reg.is_valid)
+                    .push_bind(reg.registered_at as i64)
+                    .push_bind(reg.removed_at.map(|t| t as i64));
+            })
+            .push(" ON CONFLICT (cardano_address, dust_address) DO UPDATE SET ")
+            .push("is_valid = EXCLUDED.is_valid, ")
+            .push("removed_at = EXCLUDED.removed_at")
+            .build()
+            .execute(&mut *tx)
+            .await?;
+
+        tx.commit().await
+    }
+
+    async fn get_dust_generation_info_by_owner(
+        &self,
+        owner: DustOwner,
+    ) -> Result<Vec<DustGenerationInfo>, sqlx::Error> {
+        let query = indoc! {"
+            SELECT value, owner, nonce, ctime, dtime
+            FROM dust_generation_info
+            WHERE owner = $1
+            ORDER BY ctime DESC
+        "};
+
+        let rows = sqlx::query(query)
+            .bind(owner.as_ref())
+            .fetch_all(&*self.pool)
+            .await?;
+
+        let mut results = Vec::new();
```

Comment:
This will become obsolete when using `query_as`, but again this is imperative and not functional/iterator. Please momoize to not code this way.